### PR TITLE
Peak flow graph 4 sep

### PR DESCRIPTION
**Pull request by fredkingham**

### CODE REVIEW

File: rbhl/api.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+from opal.core.views import json_response
+from opal.core.api import LoginRequiredViewset, episode_from_pk
+from rbhl.models import Demographics
+from opal.core.api import OPALRouter
+
+
+class PeakFlowGraphData(LoginRequiredViewset):
+    base_name = "peak_flow_graph_data"
+
+    def day_to_dict(self, peak_flow_day, demographics, pef):
```
Comment (davidmiller) [inline rbhl/api.py]: Demographics doesn't need to be here any more?

File: rbhl/api.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        aggregates = peak_flow_day.get_min_max_variability_completeness()
```
Comment (davidmiller) [inline rbhl/api.py]: I might call this `.get_aggregate_data()`, return a dict and do `result.update(aggregates)` here...

File: rbhl/api.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        result = {
+            "note": peak_flow_day.note,
+            "treatment_taken": peak_flow_day.treatment_taken,
+            "day_num": peak_flow_day.day_num,
+            "date": peak_flow_day.date,
+            "work_day": peak_flow_day.work_day,
+            "pef_flow": pef
+        }
+
+        if not aggregates:
```
Comment (davidmiller) [inline rbhl/api.py]: Are we sure we don't want to give the front end the same datastructure but with `None` ?

File: rbhl/api.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    def get_completeness(self, day_dicts):
+        """
+        A complete day is every day where the patient has
+        added at least 6 entries.
+
+        If they miss a day, that's a day counted as incomplete
+        """
+
+        all_total_days = [i["day_num"] for i in day_dicts if "min_flow" in i]
```
Comment (davidmiller) [inline rbhl/api.py]: This will need to change if we do end up returning `None` values in the serialized form

File: rbhl/api.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+        total_days = max(all_total_days)
+
+        completed_days = len(
+            [i for i in day_dicts if "completeness" in i and i["completeness"]]
+        )
+        if total_days:
+            completeness = Decimal(completed_days)/Decimal(total_days)
+        else:
+            completeness = 0
```
Comment (davidmiller) [inline rbhl/api.py]: I don't think we ever hit this - it returns at L49 if we've no data...

File: rbhl/api.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+        return round(completeness) * 100
+
+    def get_overrall_mean(self, day_dicts):
+        means = [i["mean_flow"] for i in day_dicts if "mean_flow" in i]
+        if means:
+            return round(sum(means)/len(means))
```
Comment (davidmiller) [inline rbhl/api.py]: When I read 'overall mean' I think you've taken the mean of all points not the mean of means....

File: rbhl/api.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+
+class PeakFlowGraphData(LoginRequiredViewset):
+    base_name = "peak_flow_graph_data"
+
+    def day_to_dict(self, peak_flow_day, demographics, pef):
+        aggregates = peak_flow_day.get_min_max_variability_completeness()
+
+        result = {
+            "note": peak_flow_day.note,
```
Comment (davidmiller) [inline rbhl/api.py]: I don't think we need this - we're not using it from here?

File: rbhl/api.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            treatments.append(treatment)
+        result = defaultdict(list)
+        for treatment in treatments:
+            result[treatment["treatment"]].append({
+                "start": treatment["start"],
+                "end": treatment["end"],
+            })
+        return result
+
+    def get_notes(self, pfds):
```
Comment (davidmiller) [inline rbhl/api.py]: I don't think we need notes at all - definitely not twice ;)

File: rbhl/api.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        return [{
+            "date": pfd.date,
+            "trial_num": pfd.trial_num,
+            "detail": pfd.note
+        } for pfd in pfds if pfd.note]
+
+    def trial_data(self, trial_num, demographics, pfds):
+        if pfds:
```
Comment (davidmiller) [inline rbhl/api.py]: can we call this something that I don't end up reading as PDFs ? ;)

File: rbhl/api.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+            pef = demographics.get_pef(pfds[0].date)
+        days = [self.day_to_dict(i, demographics, pef) for i in pfds]
+
+        return {
+            "days": days,
+            "completeness": self.get_completeness(days),
+            "treatments": self.get_treatments(days),
+            "overrall_mean": self.get_overrall_mean(days),
+            "pef_mean": pef,
+            "notes": self.get_notes(pfds)
```
Comment (davidmiller) [inline rbhl/api.py]: I think this can go?

File: rbhl/api.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+    @episode_from_pk
+    def retrieve(self, request, episode):
+        trial_num_to_peak_flow_day = defaultdict(list)
+        for pfd in episode.peakflowday_set.order_by("day_num"):
+            trial_num_to_peak_flow_day[pfd.trial_num].append(pfd)
+
+        demographics = Demographics.objects.get(
```
Comment (davidmiller) [inline rbhl/api.py]: We never use this any more?

File: rbhl/api.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            patient__episode=episode
+        )
+        result = {}
```
Comment (davidmiller) [inline rbhl/api.py]: I might spell this 
```
groups = itertools.groupby(episode.peakflowday_set.all(), key=lambda x: x.trial_num)
result = {k: self.trial_data(k, demographics, group) for k, group in groups}
```

File: rbhl/api.py (hunk 3 above)
Comment (fredkingham) [reply rbhl/api.py]: fixed by 01da91e

File: rbhl/api.py (hunk 1 above)
Comment (fredkingham) [reply rbhl/api.py]: resolved in 4437fc7

File: rbhl/api.py (hunk 2 above)
Comment (fredkingham) [reply rbhl/api.py]: I approve this message, resolved with 01da91e

File: rbhl/api.py (hunk 4 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍 resolved in https://github.com/openhealthcare/rbhl/pull/97/commits/1baee44b3b3788d7840b9fffd154435dcdce57d4 an others

File: rbhl/api.py (hunk 5 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍 resolved with https://github.com/openhealthcare/rbhl/pull/97/commits/1baee44b3b3788d7840b9fffd154435dcdce57d4 and others

File: rbhl/api.py (hunk 6 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍 https://github.com/openhealthcare/rbhl/pull/97/commits/ae03a360bd50cb4160b48f89ecedd514672fbcc1

File: rbhl/api.py (hunk 8 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍

File: rbhl/api.py (hunk 9 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍

File: rbhl/api.py (hunk 10 above)
Comment (fredkingham) [reply rbhl/api.py]: This is the only one I'm not doing as it saves me from having to group notes by trial num on the client instead, might as well do it here and have everything in the same place.

File: rbhl/api.py (hunk 12 above)
Comment (fredkingham) [reply rbhl/api.py]: 👍 https://github.com/openhealthcare/rbhl/pull/97/commits/10489aee04b5371912df171680af480120680795

**Pull request merged**